pkg/controller/sandbox/sandbox_controller.go

```diff
@@ -91,45 +91,27 @@ func (r *SandboxReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 	// Fetch the sandbox instance
 	box := &agentsv1alpha1.Sandbox{}
 	err := r.Get(ctx, req.NamespacedName, box)
-
 	if err != nil {
 		return reconcile.Result{}, client.IgnoreNotFound(err)
-	} else if box.Status.Phase == agentsv1alpha1.SandboxFailed || box.Status.Phase == agentsv1alpha1.SandboxSucceeded {
-		return reconcile.Result{}, nil
 	}
 	logger := logf.FromContext(ctx).WithValues("sandbox", klog.KObj(box))
 
 	if box.Spec.Template == nil {
+		logger.Info("sandbox template is nil, and ignore")
 		return reconcile.Result{}, nil
 	}
 
-	// Add finalizer
-	if box.DeletionTimestamp.IsZero() && !controllerutil.ContainsFinalizer(box, utils.SandboxFinalizer) {
-		err = utils.PatchFinalizer(r.Client, box, utils.AddFinalizerOpType, utils.SandboxFinalizer)
-		if err != nil {
-			logger.Error(err, "update sandbox finalizer failed")
-			return reconcile.Result{}, err
-		}
-		logger.Info("add sandbox finalizer success")
-	}
-
-	// Check Shutdown
-	now := metav1.Now()
-	var requeueAfter time.Duration
-	if box.Spec.ShutdownTime != nil && box.DeletionTimestamp == nil {
-		if box.Spec.ShutdownTime.Before(&now) {
-			logger.Info("Sandbox is shutdown time reached, will be deleted")
-			return ctrl.Result{}, r.Delete(ctx, box)
-		}
-		requeueAfter = box.Spec.ShutdownTime.Sub(now.Time)
-	}
-
 	logger.V(consts.DebugLogLevel).Info("Began to process Sandbox for reconcile")
 	newStatus := box.Status.DeepCopy()
+	hash, hashWithoutImageAndResource := core.HashSandbox(box)
```

Comment:
consider add a func `calculateStatus` and all status related calculation inside calculateStatus